src/main/java/org/tarantool/TarantoolClientImpl.java

```diff
@@ -602,27 +659,120 @@ public void close() {
 
     }
 
-    protected class ComposableAsyncOps
-        extends AbstractTarantoolOps<Integer, List<?>, Object, CompletionStage<List<?>>> {
+    protected boolean isDead(TarantoolOp<?> future) {
+        if (this.thumbstone != null) {
+            fail(future, new CommunicationException("Connection is dead", thumbstone));
+            return true;
+        }
+        return false;
+    }
 
-        @Override
-        public CompletionStage<List<?>> exec(Code code, Object... args) {
-            return (CompletionStage<List<?>>) TarantoolClientImpl.this.doExec(code, args);
+    protected static class TarantoolOp<V> extends CompletableFuture<V> {
+
+        /**
+         * A task identifier used in {@link TarantoolClientImpl#futures}.
+         */
+        private final long id;
+
+        /**
+         * Tarantool binary protocol operation code.
+         */
+        private final Code code;
+
+        /**
+         * Arguments of operation.
+         */
+        private final Object[] args;
+
+        public TarantoolOp(long id, Code code, Object[] args) {
+            this.id = id;
+            this.code = code;
+            this.args = args;
```

Comment:
I merged ExpirableOp and TarantoolOp because an _expirable feature_ was pulled up to the `TarantoolClientImpl` and reused in a cluster implementation (ExpirableOp with extra properties looked redundant). Now we have full operation-aware future that can be used for internal purposes. It doesn't impact on performance.

Reply:
Okay.